src/main/java/com/urbanairship/datacube/dbharnesses/HBaseDbHarness.java

```diff
@@ -322,43 +393,79 @@ private void flushBatch(Batch<T> batch) throws IOException, InterruptedException
-                final long nanoTimeBeforeWrite = System.nanoTime();
-
-                byte[] dbBytes;
-                switch (commitType) {
-                    case INCREMENT:
-                        long postIncr = increment(rowKey, op);
-                        dbBytes = Bytes.toBytes(postIncr);
-                        break;
-                    case READ_COMBINE_CAS:
-                        dbBytes = readCombineCas(rowKey, op);
-                        break;
-                    case OVERWRITE:
-                        overwrite(rowKey, op);
-                        dbBytes = op.serialize();
-                        break;
-                    default:
-                        throw new RuntimeException("Unsupported commit type " + commitType);
+            if (commitType.equals(CommitType.INCREMENT) && configuration.batchSize > 1) {
+                Iterable<List<Map.Entry<Address, T>>> partitions = Iterables.partition(batchMap.entrySet(), configuration.batchSize);
+                int batchesPerFlush = 0;
+                Map<BoxedByteArray, T> increments = new HashMap<>();
+                Map<BoxedByteArray, Address> backwards = new HashMap<>();
+                for (List<Map.Entry<Address, T>> entries : partitions) {
+                    batchesPerFlush++;
+                    final long nanoTimeBeforeWrite = System.nanoTime();
+                    for (Map.Entry<Address, T> entry : entries) {
+                        final Address address = entry.getKey();
+                        final T op = entry.getValue();
+                        final BoxedByteArray rowKey = new BoxedByteArray(ArrayUtils.addAll(uniqueCubeName, address.toWriteKey(idService)));
+                        increments.put(rowKey, op);
+                        backwards.put(rowKey, address);
+                    }
+
+                    Map<BoxedByteArray, byte[]> successes = increment(increments);
+
+                    for (Map.Entry<BoxedByteArray, byte[]> entry : successes.entrySet()) {
+                        successfulAddresses.add(backwards.get(entry.getKey()));
+                        successfulRows.put(entry.getKey().bytes, entry.getValue());
+                    }
+
+                    long writeDurationNanos = System.nanoTime() - nanoTimeBeforeWrite;
+                    batchWritesTimer.update(writeDurationNanos, TimeUnit.NANOSECONDS);
+                }
+                batchesPerFlushHisto.update(batchesPerFlush);
+
+                if (successfulAddresses.size() < batchMap.size()) {
+                    // the implementation prior to the addition of the batch increment code assumes any failed increment
+                    // operation results in an io exception. This matches that expectation.
+                    incrementFailuresPerFlush.update(batchMap.size() - successfulAddresses.size());
+                    throw new IOException("Some writes failed; queueing retry");
```

Comment:
Minor but might be helpful to log how many failures vs success

Reply:
https://github.com/urbanairship/datacube/pull/86/commits/9d30f8942dbd252d521548cd644fb21b95521240